### ISSUE TITLE
numpy.load: OSError: first argument must be an open file

### ISSUE THREAD

**Issue by jor-** [numpy/numpy#6766]
Using Python 3.5 and Numpy 1.10.1 I got an error at `numpy.load` when called on an opened file object. 
```
file = 'xyz.npy'
with open(file, 'rb') as file_object:
    return numpy.load(file_object)
```
The error is thrown at: numpy/lib/format.py", line 649, in read_array
    array = numpy.fromfile(fp, dtype=dtype, count=count)
OSError: first argument must be an open file

I checked with pdb: The file was closed somewhere inside the `numpy.load` call.

Unfortunately, the error does not always occur. I repeat loading the file many times in my script, the error occurs normally after several hundred loadings.

Does anyone have an explanation of why it is sometimes closed?

**Comment by jor-** [comment]
The error also occurred (after several hundred loadings) if I use `numpy.load` with the filename instead of the file object.

```
file = 'xyz.npy'
return numpy.load(file)
```

**Comment by charris** [comment]
Hmm, master works for Python3.4 and no recent changes. What platform are you on?

**Comment by jor-** [comment]
I am on Red Hat Enterprise Linux Server release 6.4 with Linux 2.6.32-358.18.1.el6.x86_64.

Maybe the problem is somehow related to multiprocessing: I load the file content in several processes simultaneously. Actually, this should not be an issue. But I do not know whether one process can close the file descriptor of another process (possibly due to a bug).

**Comment by pv** [comment]
Can you give a reproducible code example (e.g. paste in gist.github.com) of how exactly you are using the files, so that it's possible to see e.g. if you are relying on inherited file handles.

**Comment by jor-** [comment]
Simplified it is something like this:
```
import numpy as np
for i in range(10000):
    a = np.load('file.npy')
    a.sum() * i # do some calculations with a
 ```
The code runs in multiple python instances simultaneously. The file is accessed using NFS.

**Comment by pv** [comment]
These multiple instances are launched via multiprocessing process pool? Does the npy file contain pickled data, etc? A self-contained repro would be useful here.

**Comment by jor-** [comment]
The multiple instances are launched by multiple starts of the python interpreter. The npy file was created using `numpy.save('file.npy', a)`. Regarding NFS, I will do some tests if the error occurs without NFS.

**Comment by jor-** [comment]
Without NFS the same error occurs. So it seems not to be related to NFS.

**Comment by pv** [comment]
Can you prepare a minimal standalone program that reproduces it?
The above for loop when run in parallel processes does not reproduce the issue here.

**Comment by jor-** [comment]
To create a minimal standalone program is difficult, since the bug is difficult to reproduce and I do not now its cause. Maybe it is somehow related to the handling of file descriptors at the operating system or in python itself.
I have rewritten my program to avoid this heavy file access and switched to Python 3.5.1. Now everything works fine. Although the reason for the failure was not found, I am grateful for the help and I am okay with closing this issue.

**Comment by jor-** [comment]
It leaves me no rest. I'll try to write a minimal standalone program later in the day.

**Comment by jor-** [comment]
I have written a minimal standalone program where the bug occurs:
https://github.com/jor-/numpy-issue-6766.git
Just check it out and start the `run.sh`. The error occurs after one or two seconds.

**Comment by jor-** [comment]
On Python 3.5.1 I got the following error on some of the processes.
```
Traceback (most recent call last):
  File "run.py", line 6, in <module>
    a = fl.load()
  File "/tmp/numpy-issue-6766/filelock/unix.py", line 211, in load
    value = self._load(self.file)
  File "/tmp/numpy-issue-6766/filelock/np.py", line 10, in _load
    return np.load(file_object)
  File "/sfs/fs5/home-sh/sunip229/local/lib/python3.5/site-packages/numpy/lib/npyio.py", line 406, in load
    pickle_kwargs=pickle_kwargs)
  File "/sfs/fs5/home-sh/sunip229/local/lib/python3.5/site-packages/numpy/lib/format.py", line 649, in read_array
    array = numpy.fromfile(fp, dtype=dtype, count=count)
OSError: first argument must be an open file
```

**Comment by jor-** [comment]
On Python 3.4.3 with Ubuntu 15.10 I got the following error on some of the processes.
```
Traceback (most recent call last):
  File "run.py", line 6, in <module>
  File "/tmp/numpy-issue-6766/filelock/unix.py", line 211, in load
  File "/tmp/numpy-issue-6766/filelock/np.py", line 10, in _load
  File "/usr/lib/python3/dist-packages/numpy/lib/npyio.py", line 394, in load
  File "/usr/lib/python3/dist-packages/numpy/lib/format.py", line 450, in read_array
OSError: first argument must be an open file
Error in sys.excepthook:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/apport_python_hook.py", line 63, in apport_excepthook
  File "/usr/lib/python3/dist-packages/apport/__init__.py", line 5, in <module>
  File "/usr/lib/python3/dist-packages/apport/report.py", line 13, in <module>
  File "<frozen importlib._bootstrap>", line 2237, in _find_and_load
  File "<frozen importlib._bootstrap>", line 2226, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1200, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1129, in _exec
  File "<frozen importlib._bootstrap>", line 1467, in exec_module
  File "<frozen importlib._bootstrap>", line 1571, in get_code
  File "<frozen importlib._bootstrap>", line 1623, in get_data
OSError: [Errno 24] Too many open files: '/usr/lib/python3.4/glob.py'

Original exception was:
Traceback (most recent call last):
  File "run.py", line 6, in <module>
  File "/tmp/numpy-issue-6766/filelock/unix.py", line 211, in load
  File "/tmp/numpy-issue-6766/filelock/np.py", line 10, in _load
  File "/usr/lib/python3/dist-packages/numpy/lib/npyio.py", line 394, in load
  File "/usr/lib/python3/dist-packages/numpy/lib/format.py", line 450, in read_array
OSError: first argument must be an open file
```

**Comment by seberg** [comment]
I don't know, but frankly it sounds like the `fd` gets closed somehow, and I doubt that numpy has anything to do with it, maybe timeouts on the NFS because you are doing so much at the same time or something? Can you check `fd.errors` afterwards, not sure but I think that might tell you at least something?

**Comment by seberg** [comment]
Nvm. seems that only exists on python 2, was only guessing anyway.

**Comment by jor-** [comment]
I checked the `errno` of the `OSError`: It is None.

This bug is not related to NFS. It also occurs on my local ext4 file system.

**Comment by seberg** [comment]
I replaced the numpy load with
```                file_object.read(50)
                file_object.read(50)
                file_object.read(50)
                file_object.read(50)
                file_object.read(50)
                file_object.read(50)
                return file_object.read()```
as expected, you get similar "too many open files" errors, maybe there are some file descriptors not being closed properly somewhere?

**Comment by seberg** [comment]
Yup.... you have a bug:
```
        ## remove
        if remove:
            try:
                self._acquire(exclusive=True, timeout=0)
            except filelock.general.FileLockTimeoutError:
                os.close(self._fd)  # Maybe here?
            except FileNotFoundError:
                pass
            else:
                try:
                    os.remove(self.lockfile)
                except OSError as e:
                    if e.errno == errno.EBUSY:
                        pass
                    else:
                        raise
                else:
                    pass
                finally:
                    self._release(remove=False)
```
I had put an extra finally with `if self._fd is not None` and close in that case and it ran fine, I expect the right place is in the `except filelock.general.FileLockTimeoutError:`

**Comment by seberg** [comment]
At least it seems to run fine with that for me.

**Comment by seberg** [comment]
Anyway, I will close this, since I am quite sure it has nothing to do with `np.load` itself, feel free to continue the discussion.

**Comment by pv** [comment]
The error message from fromfile was confusing, but that was fixed in 5225e4c2007963f

**Comment by jor-** [comment]
@seberg thank you very much for the hint. You saved my day.

Nevertheless the error message is very misleading especially with Python 3.5. Maybe it makes sense to change the error message at the appropriate location.

**Comment by jor-** [comment]
Okay, then I'm happy, and thank everyone for their help.

**Comment by pv** [comment]
The error message has been fixed some months ago.
